spyder-ide/docrepr
# Update to work with latest Sphinx and IPython

Per @ccordoba12's comment in #19, here's a PR that updates this package to work with current Sphinx and IPython.

There were three changes required:
1. Sphinx no longer allows input and output directories to be the same. We create a temp directory to use as the output.
2. `oinspect` in IPython no longer has `argspec`, so we check for its existence.
3. The jinja2 template needs to check whether there's an actual `warning_message` before rendering it, not just the existence of a `warning` key.

To test, first set up a new conda env:
```console
conda create -n sphinx-docrepr-stuff-issue
conda activate sphinx-docrepr-stuff-issue
conda install -c conda-forge sphinx numpy ipython
```

Then, run the following in this branch:
```py
import docrepr                                # Set module options
from docrepr import sphinxify                 # html generator
from IPython.core.oinspect import Inspector   # oinfo generator

import numpy as np

oinfo = Inspector().info(np.sin)
oinfo['name'] = 'sin'
url = sphinxify.rich_repr(oinfo)
print('file://' + url)
```

> Of course, you can also check out this branch and run `python -m pip install .` to install this branch of docrepr, then run the above snippet from any directory.

This produces the following:

<img width="963" alt="Screen Shot 2021-11-16 at 9 53 15 AM" src="https://user-images.githubusercontent.com/37649/142039172-1f682e53-9a5b-4315-9f56-5fffd093773c.png">

With this installed, we can also confirm Jupyter and JupyterLab formats `?` docstrings out of the box! Launch `jupyter lab`, create a cell with
```py
ip = get_ipython()
ip.sphinxify_docstring = True
ip.enable_html_pager = True

import numpy
```
evaluate this, and in another cell, enter and run
```
numpy.sin?
```

![docrepr-notebook](https://user-images.githubusercontent.com/37649/142039919-3990b2de-36c5-4af1-85b9-e463a8d0f1ee.gif)

This leverages functionality added by @SylvainCorlay (see [provisional changes](https://github.com/ipython/ipython/blob/master/docs/source/whatsnew/version5.rst#provisional-changes)) in IPython (and thus ipykernel) to take advantage of docrepr if (1) it's installed and (2) if `sphinxify_docstring` is true, as we set above.

> **Nota bene** Docrepr isn't used in the JupyterLab *contextual help* yet. That requires this one-line PR in IPython: https://github.com/ipython/ipython/pull/13290 😄 if that lands, then docrepr can format documentation *everywhere* in JupyterLab! Keeping my fingers crossed.
